### ISSUE TITLE
Trouble getting lidar_map_api or tgc_image_terrain to run

### ISSUE THREAD

**Issue by frodaddy00** [chadrockey/TGC-Designer-Tools#3]
Hello! I'm very excited about this project. I am having problems trying to run lidar_map_gui or tgc_image_terrain. I can run tgc_gui but I don't really know how these UI screens work yet!

I pulled down your source and loaded into pycharm. I am running Python 3.7.2. I've tried a couple of different sets of LAS files but here is the one that got me the farthest.

https://chs.coast.noaa.gov/pickup/dav//296650/Job452277_ga2010_arc_dekalbcounty.zip
Lidar data file:                Job452277_ga2010_arc_dekalbcounty.las
Metadata file:                  Job452277_ga2010_arc_dekalbcounty_metadata.xml

I created a folder and put the .las and .xml file in it so its at least trying to read it.

****
lidar_map.api gives me this error..

C:\projects\TGC-Designer-Tools-master\venv\Scripts\python.exe C:/projects/TGC-Designer-Tools-master/lidar_map_api.py c:\lidar
Processing: Job452277_ga2010_arc_dekalbcounty.las
Using metadata: Job452277_ga2010_arc_dekalbcounty_metadata.xml
Unknown unit: Unknown.  Can't process data accurately, please report an issue and send this lidar and metadata
Unit in metadata is 1.0
Lidar image sampled size: 3342 x 4005
Generating brightness and height images
0% visualizing lidar
17% visualizing lidar
38% visualizing lidar
59% visualizing lidar
80% visualizing lidar
Adding golf features to lidar data
Traceback (most recent call last):
  File "C:/projects/TGC-Designer-Tools-master/lidar_map_api.py", line 137, in <module>
    upper_left_latlon = pc.enuToLatLon(*upper_left_enu)
  File "C:\projects\TGC-Designer-Tools-master\GeoPointCloud.py", line 134, in enuToLatLon
    return self.projToLatLon(proj[0], proj[1])
  File "C:\projects\TGC-Designer-Tools-master\GeoPointCloud.py", line 129, in projToLatLon
    lonlat = pyproj.transform(self._proj, self._platlon, easting, northing)
  File "C:\projects\TGC-Designer-Tools-master\venv\lib\site-packages\pyproj\__init__.py", line 515, in transform
    raise TypeError("p1 must be a Proj class")
TypeError: p1 must be a Proj class

Process finished with exit code 1

******
tgc_image_terrain gives me this error:

C:\projects\TGC-Designer-Tools-master\venv\Scripts\python.exe C:/projects/TGC-Designer-Tools-master/tgc_image_terrain.py c:\lidar
Loading data
Filling holes in heightmap
Traceback (most recent call last):
  File "C:/projects/TGC-Designer-Tools-master/tgc_image_terrain.py", line 49, in <module>
    read_dictionary = np.load(lidar_dir_path + '/output/heightmap.npy').item()
  File "C:\projects\TGC-Designer-Tools-master\venv\lib\site-packages\numpy\lib\npyio.py", line 415, in load
    fid = open(os_fspath(file), "rb")
FileNotFoundError: [Errno 2] No such file or directory: 'c:\\lidar/output/heightmap.npy'

Process finished with exit code 1

***
Finally here is my pycrs and package info...

Name: PyCRS
Version: 1.0.0
Summary: GIS package for reading, writing, and converting between CRS formats.
Home-page: http://github.com/karimbahgat/PyCRS
Author: Karim Bahgat
Author-email: karim.bahgat.norway@gmail.com
License: MIT
Location: c:\users\fro\appdata\local\programs\python\python37\lib\site-packages
Requires:
Required-by:

Package         Version
--------------- -------
cycler          0.10.0
kiwisolver      1.0.1
laspy           1.5.1
matplotlib      3.0.2
numpy           1.16.1
Pillow          5.4.1
pip             19.0.1
PyCRS           1.0.0
pyparsing       2.3.1
pyproj          1.9.6
python-dateutil 2.8.0
PyYAML          3.13
scipy           1.2.0
setuptools      40.6.2
six             1.12.0

**Comment by chadrockey** [comment]
Thanks.  I'll need to make that unit message more prominent and probably fail when the geo projection can't be determined.

The file you linked is new to me and only has an EPSG code.  We might be able to allow users to overwrite the projection with EPSG codes if they can look through various formats and find them.  I think it's a losing battle to support every format unless we can find a library that can handle them.

That being said, it's clearly possible for some programs to find the right data, so maybe we can do it too.

I use NOAA and then usually the National Map to get the lidar:
https://coast.noaa.gov/inventory/
https://viewer.nationalmap.gov/basic/

I found your data in the USGS metadata format I've been using otherwise:
https://www.sciencebase.gov/catalog/item/58278d07e4b01fad8703fd70
https://www.sciencebase.gov/catalog/item/58278d11e4b01fad8703fe5b
https://www.sciencebase.gov/catalog/item/58278d04e4b01fad8703fc92
https://www.sciencebase.gov/catalog/item/58278d07e4b01fad8703fd78

Maybe start with these four files, just so you can keep moving for now?

![smokerise](https://user-images.githubusercontent.com/230240/52470383-d56c6f00-2b42-11e9-96f1-120404dec150.PNG)

Of course, my file isn't even processing this right so there's no sat preview!  xD  I'll sort this out more tomorrow.

**Comment by frodaddy00** [comment]
I did start with those two links you mentioned but apparently was stumbling on the nationalmap website and did not select the proper lidar files.

https://coast.noaa.gov/inventory/
https://viewer.nationalmap.gov/basic/

I can see now that you can draw a box around the course (Smoke Rise Country Club) and it ended up with 6 different pointcloud files. Will it combine these if I add them to cart? Here is what it looks like.

![capture3](https://user-images.githubusercontent.com/47437707/52487927-be526f00-2b8c-11e9-9158-1992ad1440fa.JPG)

Overall I suspect getting the appropriate lidar files will be the biggest learning curve for users.

But yes if you have files that work I'm more than happy to use those to finish my course!

**Comment by chadrockey** [comment]
Wow!  I never knew you could draw a box!  I'll have to see what the cart feature is as well.

What I've been doing is clicking on footprint one by one to make sure the whole course was in view, then clicking on Info/Metadata in a new tab and then downloading files from each tab.

This is great and hopefully we can find a way to streamline this.  I really wanted to support automatic lidar grabbing: you enter your course name or address and the software fetches it, but I think you have to maintain your own lidar database/server to do so.

As for the multiple lidar question, yes, the program will do its best to combine all lidar files in the directory you provide, even if they are from different years or have different geographic projections.  I don't recommend mixing lidar years as I had bad results.  For 6 files from the same year like you have, that's an intended usecase.

**Comment by chadrockey** [comment]
I dove down the projection rabbit hole again and added more spaghetti code.  In general, I'm going to try to lookup and use the EPSG code when possible.

Fixed in https://github.com/chadrockey/TGC-Designer-Tools/commit/c760a954a70ac4edf3d161a2b8f9876100238259
I made it so that both sets of lidar data for your course are functional:

![workingnoaa](https://user-images.githubusercontent.com/230240/52517898-77e62a00-2bf7-11e9-91be-790f1fe048dd.PNG)

![workingusgs](https://user-images.githubusercontent.com/230240/52517899-7b79b100-2bf7-11e9-9d6d-7c8240809f56.PNG)

I'll consider this issue scoped to the specific lidar metadata.  Let's open a separate issue for future problems.
